modules/abstract-eth/src/abstractEthLikeNewCoins.ts

```diff
@@ -1926,35 +1930,100 @@ export abstract class AbstractEthLikeNewCoins extends AbstractEthLikeCoin {
+          params.walletPassphrase
+        );
+        const backupKeyPair = new KeyPairLib({ pub: backupKeyCombined.xShare.y });
+        const baseAddress = backupKeyPair.getAddress();
 
-      let { tx } = await this.buildTssRecoveryTxn(baseAddress, gasPrice, gasLimit, params);
+        unsignedTx = (await this.buildTssRecoveryTxn(baseAddress, gasPrice, gasLimit, params)).tx;
 
-      const signableHex = tx.getMessageToSign(false).toString('hex');
-      if (!userKeyCombined || !backupKeyCombined) {
-        throw new Error('Missing key combined shares for user or backup');
+        signature = await this.signRecoveryTSS(
+          userKeyCombined,
+          backupKeyCombined,
+          unsignedTx.getMessageToSign(false).toString('hex')
+        );
+      } else {
+        const { userKeyShare, backupKeyShare, commonKeyChain, baseAddress } = this.getMpcV2RecoveryKeyShares(
+          userPublicOrPrivateKeyShare,
+          backupPrivateOrPublicKeyShare,
+          params.walletPassphrase
+        );
+
+        unsignedTx = (await this.buildTssRecoveryTxn(baseAddress, gasPrice, gasLimit, params)).tx;
+
+        signature = await AbstractEthLikeNewCoins.signRecoveryMpcV2(
+          unsignedTx,
+          userKeyShare,
+          backupKeyShare,
+          commonKeyChain
+        );
       }
-      const signature = await this.signRecoveryTSS(userKeyCombined, backupKeyCombined, signableHex);
+
       const ethCommmon = AbstractEthLikeNewCoins.getEthLikeCommon(params.eip1559, params.replayProtectionOptions);
-      tx = this.getSignedTxFromSignature(ethCommmon, tx, signature);
+      const signedTx = this.getSignedTxFromSignature(ethCommmon, unsignedTx, signature);
 
       return {
-        id: addHexPrefix(tx.hash().toString('hex')),
-        tx: addHexPrefix(tx.serialize().toString('hex')),
+        id: addHexPrefix(signedTx.hash().toString('hex')),
+        tx: addHexPrefix(signedTx.serialize().toString('hex')),
       };
-    } else {
-      // DKLS
-      throw new Error('DKLS recovery is not implemented yet');
     }
   }
 
+  private static async signRecoveryMpcV2(
+    tx: EthLikeTxLib.FeeMarketEIP1559Transaction | EthLikeTxLib.Transaction,
+    userKeyShare: string,
+    backupKeyShare: string,
+    commonKeyChain: string
+  ) {
+    const messageHash = tx.getMessageToSign(true);
+    const userDsg = new DklsDsg.Dsg(Buffer.from(userKeyShare, 'base64'), 0, 'm', messageHash);
+    const backupDsg = new DklsDsg.Dsg(Buffer.from(backupKeyShare, 'base64'), 1, 'm', messageHash);
+
+    const signatureString = DklsUtils.verifyAndConvertDklsSignature(
+      messageHash,
+      (await DklsUtils.executeTillRound(5, userDsg, backupDsg)) as DklsTypes.DeserializedDklsSignature,
+      commonKeyChain,
+      'm',
+      undefined,
+      false
+    );
+    const sigParts = signatureString.split(':');
+
+    return {
+      recid: parseInt(sigParts[0], 10),
+      r: sigParts[1],
+      s: sigParts[2],
+      y: sigParts[3],
+    };
+  }
```

Comment:
Dont we have a re-usable function from gg18 tss for this?

Reply:
I didn't find it in the utils